### DISCUSSION TITLE
Support `partial properties` from C#13

### DISCUSSION THREAD

**Discussion by michalciolek** [canton7/PropertyChanged.SourceGenerator#44] [Ideas]
Microsoft has just [released a new version of .NET 9 Preview 6](https://devblogs.microsoft.com/dotnet/dotnet-9-preview-6/), which includes [partial properties](https://devblogs.microsoft.com/dotnet/csharp-13-explore-preview-features/#partial-properties) in C# 13. Are there any plans to add support for this feature in source generators?

**Comment by canton7** [comment]
Yep, I was excited to see that announced out of the blue! The plan is to support them - I've been trying to push the benefits of partial properties for years. Currently undecided whether to support them and the current field based approach together, or just do a clean break in a new major version.

**Reply by michalciolek** [reply]
When creating the new version, could you please consider ensuring an easy migration path from version 1.0? Specifically, it would be helpful to support gradual updates and the replacement of fields with partial properties. Additionally, for users migrating from the Fody library, simply adding `partial` to properties should suffice (without [Notify] attribute).

In my opinion, it would also be a good idea to release version 2.0 with partial property support, which drops support for .NET Framework and .NET 8 and earlier, while retaining the mechanisms present in version 1.0.

PS. Thank you for your work on this project.

**Reply by canton7** [reply]
> Additionally, for users migrating from the Fody library, simply adding partial to properties should suffice (without [Notify] attribute).

This is not possible for multiple reasons.

Source Generators need to be fast, and there are built-in mechanisms for responding to changes on symbols decorated with a particular attribute. Not using these would result in unacceptable slowness.

Also, I need to know which properties to generate partial implementations for. If I generated partial implementations for all partial properties, then you would see compilation errors if any other source generator also tried to generate implementations for those partial properties. This means that you would be unable to use `[GeneratedRegex]` etc anywhere in your assembly.

> When creating the new version, could you please consider ensuring an easy migration path from version 1.0? Specifically, it would be helpful to support gradual updates and the replacement of fields with partial properties.

I wondered about that, but it's tricky in practice. Quite a lot of fairly fundamental logic changes with the move to partial properties, and running the two code paths side-by-side is going to be hard. Also, most of the parameters to the `Notify` attribute aren't relevant to partial properties.

That's not to say it's not do-able, but the amount of effort means that it I might never have time to release support for partial properties at all!

There might be an alternative route to automatically upgrade from the field-based approach to the partial property approach.

> In my opinion, it would also be a good idea to release version 2.0 with partial property support, which drops support for .NET Framework and .NET 8 and earlier,

Why? The target framework isn't relevant here -- what matters is whether the compiler version supports partial properties.
